integration/test_notifications.py

```diff
@@ -0,0 +1,122 @@
+"""Test notfiications on peer login and logout."""
+
+import asyncio
+import logging
+import os
+from typing import Callable
+
+import pytest
+import websockets.client as ws
+from olm import Account
+
+from client import Client
+from client.session_manager_client import SessionManager
+from common.keys import KeyPair, PubKey
+from common.messages import CansMessage, CansMsgId
+
+
+class NotificationsSuccess(Exception):
+    """Dummy exception raised to gracefully exit the event loop."""
+
+    ...
+
+
+class MockSessionManager(SessionManager):
+    """Mock session manager."""
+
+    async def _handle_upstream(self, conn: ws.WebSocketClientProtocol) -> None:
+        """Mock upstream handler."""
+        self.log.debug("Mock upstream handler called at session manager level")
+
+    async def _handle_downstream(
+        self, conn: ws.WebSocketClientProtocol
+    ) -> None:
+        """Mock downstream handler."""
+        self.log.debug(
+            "Mock downstream handler called at session manager level"
+        )
+
+
+class MockClient(Client):
+    """Mock client."""
+
+    def __init__(
+        self,
+        my_keys: KeyPair,
+        peer_pub_key: PubKey,
+        session_manager_factory: Callable,
+    ) -> None:
+        """Construct mock client."""
+        self.server_hostname = os.environ["CANS_SERVER_HOSTNAME"]
+        self.server_port = os.environ["CANS_PORT"]
+        self.certpath = os.environ["CANS_SELF_SIGNED_CERT_PATH"]
+        self.event_loop = asyncio.get_event_loop()
+        self._do_logger_config()
+
+        self.test_peer = peer_pub_key
+
+        self.log = logging.getLogger("cans-logger")
+        self.log.setLevel(logging.DEBUG)
+
+        account = Account()
+        self.session_manager = session_manager_factory(
```

Comment:
Naming here is not clear to me. The attribute is called `session_manager_factory` which suggests that it is a factory method creational pattern, but later we are passing classes